{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Laurent Jutras-DubÃ©\n",
    "\n",
    "## The 3-gene models: asymmetric patterns\n",
    "\n",
    "####   \n",
    "\n",
    "\n",
    "**1. Pattern formation with the asymmetric 3-gene models**\n",
    "\n",
    "In this notebook, we perform deterministic simulations of the 3-gene models. We introduce asymmetries in the repression thresholds of the static term, which leads to the formation of an asymmetric spatial pattern of protein expression. See the Methods and Supplement for a description of the models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "%matplotlib notebook\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors\n",
    "import matplotlib.patheffects as PathEffects\n",
    "plt.rcParams.update({'figure.max_open_warning': 0})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####     \n",
    "\n",
    "**1.1 The ODEs**\n",
    "\n",
    "In the following code cell, we define functions for the ODEs describing the dynamics of the 3 proteins. We make sure that we can easily modify the parameters of the repression interactions to generate asymmetries in the spatial pattern of protein expression. We specify the values of all parameters later in this notebook. Note that the parameters of the Hill function describing the repression of, say, protein $A$ by protein $B$ in the static term are named `stat_K_BA` and `stat_n_BA`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dynamic(conc, param):\n",
    "    \n",
    "    A = conc[:,0]\n",
    "    B = conc[:,1]\n",
    "    C = conc[:,2]\n",
    "    \n",
    "    K_BA = param[\"dyn_K_BA\"]\n",
    "    K_CB = param[\"dyn_K_CB\"]\n",
    "    K_AC = param[\"dyn_K_AC\"]\n",
    "    \n",
    "    n_BA = param[\"dyn_n_BA\"]\n",
    "    n_CB = param[\"dyn_n_CB\"]\n",
    "    n_AC = param[\"dyn_n_AC\"]\n",
    "       \n",
    "    dA = 1./(1.+(B/K_BA)**n_BA)\n",
    "    dB = 1./(1.+(C/K_CB)**n_CB)\n",
    "    dC = 1./(1.+(A/K_AC)**n_AC)\n",
    "    \n",
    "    return np.array([dA, dB, dC]).transpose()\n",
    "\n",
    "\n",
    "\n",
    "def static(conc, param):\n",
    "    \n",
    "    A = conc[:,0]\n",
    "    B = conc[:,1]\n",
    "    C = conc[:,2]\n",
    "    \n",
    "    K_BA = param[\"stat_K_BA\"]\n",
    "    K_CB = param[\"stat_K_CB\"]\n",
    "    K_AC = param[\"stat_K_AC\"]\n",
    "    K_CA = param[\"stat_K_CA\"]\n",
    "    K_AB = param[\"stat_K_AB\"]\n",
    "    K_BC = param[\"stat_K_BC\"]\n",
    "    \n",
    "    n_BA = param[\"stat_n_BA\"]\n",
    "    n_CB = param[\"stat_n_CB\"]\n",
    "    n_AC = param[\"stat_n_AC\"]\n",
    "    n_CA = param[\"stat_n_CA\"]\n",
    "    n_AB = param[\"stat_n_AB\"]\n",
    "    n_BC = param[\"stat_n_BC\"]\n",
    "       \n",
    "    dA = 1./(1.+(B/K_BA)**n_BA) /(1.+(C/K_CA)**n_CA)\n",
    "    dB = 1./(1.+(C/K_CB)**n_CB) /(1.+(A/K_AB)**n_AB)\n",
    "    dC = 1./(1.+(A/K_AC)**n_AC) /(1.+(B/K_BC)**n_BC)\n",
    "    \n",
    "    return np.array([dA, dB, dC]).transpose()\n",
    "\n",
    "\n",
    "\n",
    "def dconc(conc, g, param):\n",
    "    \n",
    "    g = np.array([g, g, g]).transpose()\n",
    "    ng = param[\"ng\"]\n",
    "    \n",
    "    return g**ng *dynamic(conc, param) +(1.-g)**ng *static(conc, param) -conc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####     \n",
    "\n",
    "**1.2 The morphogen gradient**\n",
    "\n",
    "In a given cell, the value of $g$ stays equal to 1 for some time (via the `x_osc` parameter) before decreasing exponentially (via the `s` parameter). To generate a spatial pattern, we make the time at which the value of $g$ starts decreasing, $t^*$, dependent on the position of the cell in the tissue: $t^*(x)$. To model a tissue growing with a constant speed, we make $t^*(x)$ a linear function of $x$, via the `positions` vector and the `v` parameter. In other words, we model a gradient of $g$ with a fixed steepness that moves in the posterior direction at aconstant rate. We generate a kymograph showing the spatio-temporal profiles of $g$ later in this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_g(t, param):\n",
    "    \n",
    "    x_osc = param[\"x_osc\"]\n",
    "    s = param[\"s\"]\n",
    "    v = param[\"v\"]\n",
    "    positions = np.linspace(0.0, 1.0, param[\"n_cells\"])\n",
    "    \n",
    "    g = np.exp(s*(positions +x_osc -v*t))\n",
    "    g[g > 1.] = 1.\n",
    "    \n",
    "    return g"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  \n",
    "\n",
    "**1.3 The integration scheme**\n",
    "\n",
    "We use the Euler integration scheme to perform the integration of the ODEs. The `integrate` function requires two inputs: the initial concentration of the 3 genes at all spatial positions, and the dictionary of parameters. Parameter `n_steps_keep_results` specifies the number of integration steps between which the simulation results are kept. Increasing this parameter can help reduce the size of the three output matrices, `results`, `g_gradient` and `times_plot`. These matrices respectively contain the concentration of the 3 genes at all positions for every time point, the spatio-temporal distribution of parameter $g$ and the array of times."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def integrate(init_conc, param):\n",
    "\n",
    "    # Set the integration parameters\n",
    "    total_t = param[\"total_t\"]\n",
    "    dt = param[\"dt\"]\n",
    "    times = np.arange(0., total_t, dt)\n",
    "\n",
    "    conc = init_conc\n",
    "    results = [conc]\n",
    "    g = compute_g(0., param)\n",
    "    g_gradient = [g]\n",
    "    times_plot = [0.]\n",
    "    \n",
    "    counter = 0\n",
    "    n_steps_keep_results = param[\"n_steps_keep_results\"]\n",
    "\n",
    "    \n",
    "    # Perform the integration\n",
    "    for t in times[1:]:\n",
    "        \n",
    "        conc = conc +dconc(conc, g, param)*dt\n",
    "        conc[conc < 0.] = 0.\n",
    "        \n",
    "        g = compute_g(t, param)\n",
    "           \n",
    "        counter += 1\n",
    "        if (counter %n_steps_keep_results == 0):\n",
    "            results.append(conc)\n",
    "            g_gradient.append(g)\n",
    "            times_plot.append(t)\n",
    "        \n",
    "\n",
    "    # Output the results\n",
    "    results = np.array(results)\n",
    "    g_gradient = np.array(g_gradient)\n",
    "    times_plot = np.array(times_plot)\n",
    "    \n",
    "    return results, g_gradient, times_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  \n",
    "\n",
    "**1.4 The kymograph plots**\n",
    "\n",
    "We visualize the results with a kymograph where the $x$ axis is time and the $y$ axis is the position of the cell along the tissue. We represent the concentration of the genes with a purple to blue to white color scheme, where white (blue and purple, resp.) means that gene $A$ ($B$ and $C$, resp.) is predominant in that cell at that time. A cell that contains a mix of the three genes appears grey on the kymograph.\n",
    "\n",
    "We plot lines representing the moment at which $g$ starts decreasing for the different positions along the tissue. We also plot lines for the moments at which $g$ reaches specific values, namely $0.5$ and $0.2$. To achieve this, we need to define a function `compute_x` that outputs the position of the cell that has a given value of $g$ at a given time, and a function `compute_t` that outputs the time at which $g$ has a given value at a given position along the tissue.\n",
    "\n",
    "Finally, we also define a function that plots the spatio-temporal profile of parameter $g$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_x(g, t, param):\n",
    "\n",
    "    x_osc = param[\"x_osc\"]\n",
    "    s = param[\"s\"]\n",
    "    v = param[\"v\"]\n",
    "    \n",
    "    return np.log(g)/s -x_osc +v*t\n",
    "\n",
    "\n",
    "\n",
    "def compute_t(g, x, param):\n",
    "\n",
    "    x_osc = param[\"x_osc\"]\n",
    "    s = param[\"s\"]\n",
    "    v = param[\"v\"]\n",
    "    \n",
    "    return (x_osc +x -np.log(g)/s)/v\n",
    "    \n",
    "\n",
    "    \n",
    "def plot_kymograph(results, g_gradient, times_plot, file_name, param):\n",
    "    \n",
    "    plt.rcParams[\"figure.figsize\"] = 6.4, 4.8\n",
    "    fig, ax = plt.subplots()\n",
    "    \n",
    "    # Kymograph\n",
    "    norm = colors.Normalize(vmin=0., vmax=1.)\n",
    "    ax.imshow(results[:,:,1].transpose(), norm=norm, cmap='PuBu', aspect='auto')\n",
    "    kymo = ax.imshow(results[:,:,2].transpose(), norm=norm, cmap='BuPu', aspect='auto', alpha=0.5)\n",
    "    \n",
    "    # Colorbar\n",
    "    cb = fig.colorbar(kymo, ax=ax, shrink=0.75, ticks=[0., 0.5, 1.])\n",
    "    cb.ax.tick_params(labelsize=24)\n",
    "    cb.ax.set_yticklabels(['A', 'B', 'C'])\n",
    "    \n",
    "    # Lines of constant g\n",
    "    ax.plot(param[\"n_cells\"]*compute_x(1., times_plot, param), color='k', linewidth=4)\n",
    "    ax.plot(param[\"n_cells\"]*compute_x(0.5, times_plot, param), color='k', linewidth=4)\n",
    "    ax.plot(param[\"n_cells\"]*compute_x(0.2, times_plot, param), color='k', linewidth=4)\n",
    "    \n",
    "    ax.text(compute_t(1., 0.55, param)/param[\"dt\"]-800., param[\"n_cells\"]*0.55, 'g=1', color='k', rotation=-72, fontsize=24)\n",
    "    ax.text(compute_t(0.5, 0.55, param)/param[\"dt\"]-50., param[\"n_cells\"]*0.55, 'g=0.5', rotation=-72, color='k', fontsize=24)\n",
    "    ax.text(compute_t(0.2, 0.55, param)/param[\"dt\"]-50., param[\"n_cells\"]*0.55, 'g=0.2', rotation=-72, color='k', fontsize=24)\n",
    "    \n",
    "    # Other plot options\n",
    "    ax.set_xlabel('$t$', fontsize=32)\n",
    "    ax.set_ylabel('$x$', fontsize=32)\n",
    "    ax.set_ylim([param[\"n_cells\"], 0])\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    \n",
    "    fig.show()\n",
    "    \n",
    "    \n",
    "    \n",
    "def plot_g_gradient(g_gradient, times_plot, file_name, param):\n",
    "\n",
    "    plt.rcParams[\"figure.figsize\"] = 6.4, 4.8\n",
    "    fig, ax = plt.subplots()\n",
    "\n",
    "    # Kymograph\n",
    "    norm = colors.Normalize(vmin=0., vmax=1.)\n",
    "    kymo = ax.imshow(g_gradient.transpose(), norm=norm, cmap='Greys', aspect='auto')\n",
    "    \n",
    "    # Colorbar\n",
    "    cb = fig.colorbar(kymo, ax=ax, shrink=0.75, ticks=[0., 0.5, 1.])\n",
    "    cb.ax.tick_params(labelsize=24)\n",
    "    \n",
    "    # Lines of constant g\n",
    "    ax.plot(param[\"n_cells\"]*compute_x(1., times_plot, param), color='w', linewidth=4)\n",
    "    ax.plot(param[\"n_cells\"]*compute_x(0.5, times_plot, param), color='k', linewidth=4)\n",
    "    ax.plot(param[\"n_cells\"]*compute_x(0.2, times_plot, param), color='k', linewidth=4)\n",
    "    \n",
    "    ax.text(compute_t(1., 0.55, param)/param[\"dt\"]-800., param[\"n_cells\"]*0.55, 'g=1', color='w', rotation=-72, fontsize=28)\n",
    "    ax.text(compute_t(0.5, 0.55, param)/param[\"dt\"]-50., param[\"n_cells\"]*0.55, 'g=0.5', rotation=-72, color='k', fontsize=28)\n",
    "    ax.text(compute_t(0.2, 0.55, param)/param[\"dt\"]-50., param[\"n_cells\"]*0.55, 'g=0.2', rotation=-72, color='k', fontsize=28)\n",
    "    \n",
    "    # Other plot options\n",
    "    ax.set_title('$g(x,t)$', fontsize=32)\n",
    "    ax.set_xlabel('$t$', fontsize=32)\n",
    "    ax.set_ylabel('$x$', fontsize=32)\n",
    "    ax.set_ylim([param[\"n_cells\"], 0])\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    \n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  \n",
    "\n",
    "**1.5 The dynamics of pattern formation in Models 1 and 2**\n",
    "\n",
    "We define all the required parameters for the ODEs, the morphogen gradient and the integration. We then use the `integrate` function to perform the integration and the `plot_kymograph` function to visualise the results. In Model 1, we set parameter `ng` to 2, while in Model 2, we set `ng` to 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "param ={\n",
    "    \n",
    "    # Integration parameters\n",
    "    \"total_t\"              : 60.,\n",
    "    \"dt\"                   : 0.01,\n",
    "    \"n_steps_keep_results\" : 1,\n",
    "    \"n_cells\"              : 500,\n",
    "\n",
    "    # ODE parameters\n",
    "    \"ng\" : 2,\n",
    "    \n",
    "    \"dyn_K_BA\" : 0.4,\n",
    "    \"dyn_K_CB\" : 0.4,\n",
    "    \"dyn_K_AC\" : 0.4,\n",
    "    \n",
    "    \"dyn_n_BA\" : 5,\n",
    "    \"dyn_n_CB\" : 5,\n",
    "    \"dyn_n_AC\" : 5,\n",
    "    \n",
    "    \"stat_K_BA\" : 0.4,\n",
    "    \"stat_K_CB\" : 0.4,\n",
    "    \"stat_K_AC\" : 0.4,\n",
    "    \"stat_K_CA\" : 0.4,\n",
    "    \"stat_K_AB\" : 0.4,\n",
    "    \"stat_K_BC\" : 0.4,\n",
    "    \n",
    "    \"stat_n_BA\" : 5,\n",
    "    \"stat_n_CB\" : 5,\n",
    "    \"stat_n_AC\" : 5,\n",
    "    \"stat_n_CA\" : 5,\n",
    "    \"stat_n_AB\" : 5,\n",
    "    \"stat_n_BC\" : 5,\n",
    "    \n",
    "    # Morphogen gradient parameters\n",
    "    \"x_osc\" : 0.2,\n",
    "    \"s\"     : 1.,\n",
    "    \"v\"     : 0.05\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "init_conc = np.zeros((param[\"n_cells\"], 3))\n",
    "init_conc[:,0] = 0.75"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####      \n",
    "\n",
    "**1.5.1 Weak asymmetry**\n",
    "\n",
    "We first introduce an asymmetry in the static term by decreasing the thresholds of repression of protein $A$ by proteins $B$ and $C$. We study the robustness of the patterns formed by Models 1 and 2 to decreases in the steepness of the spatial profile of parameter $g$, which can be interpreted as a morphogen concentration. Note that we model embryo elongation by moving the morphogen gradient in the posterior direction at a constant speed. Therefore, decreasing the spatial steepness of the morphogen is equivalent to decreasing the time-scale of morphogen decay in single cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "param[\"stat_K_AB\"] = 0.4\n",
    "param[\"stat_K_AC\"] = 0.4\n",
    "param[\"stat_K_BA\"] = 0.36\n",
    "param[\"stat_K_BC\"] = 0.4\n",
    "param[\"stat_K_CA\"] = 0.36\n",
    "param[\"stat_K_CB\"] = 0.4\n",
    "\n",
    "\n",
    "# Steep gradient for g\n",
    "param[\"s\"] = 2.5\n",
    "\n",
    "param[\"ng\"] = 2\n",
    "results_asym1_steep_m1, g_gradient_asym1_steep_m1, times_plot_asym1_steep_m1 = integrate(init_conc, param)\n",
    "\n",
    "param[\"ng\"] = 1\n",
    "results_asym1_steep_m2, g_gradient_asym1_steep_m2, times_plot_asym1_steep_m2 = integrate(init_conc, param)\n",
    "\n",
    "\n",
    "\n",
    "# Shallow gradient for g\n",
    "param[\"s\"] = 1.\n",
    "\n",
    "param[\"ng\"] = 2\n",
    "results_asym1_shallow_m1, g_gradient_asym1_shallow_m1, times_plot_asym1_shallow_m1 = integrate(init_conc, param)\n",
    "\n",
    "param[\"ng\"] = 1\n",
    "results_asym1_shallow_m2, g_gradient_asym1_shallow_m2, times_plot_asym1_shallow_m2 = integrate(init_conc, param)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####     \n",
    "\n",
    "In the following code cell, we plot the results for Models 1 and 2 with a steep profile of parameter $g$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "param[\"s\"] = 2.5\n",
    "\n",
    "plot_g_gradient(g_gradient_asym1_steep_m1, times_plot_asym1_steep_m1, '3gene_asym1_steep_kymo_g.png', param)\n",
    "plot_kymograph(results_asym1_steep_m1, g_gradient_asym1_steep_m1, times_plot_asym1_steep_m1, '3gene_asym1_steep_kymo_m1.png', param)\n",
    "plot_kymograph(results_asym1_steep_m2, g_gradient_asym1_steep_m2, times_plot_asym1_steep_m2, '3gene_asym1_steep_kymo_m2.png', param)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####     \n",
    "\n",
    "In the following code cell, we plot the results for Models 1 and 2 with a shallow profile of parameter $g$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "param[\"s\"] = 1.\n",
    "\n",
    "plot_g_gradient(g_gradient_asym1_shallow_m1, times_plot_asym1_shallow_m1, '3gene_asym1_shallow_kymo_g.png', param)\n",
    "plot_kymograph(results_asym1_shallow_m1, g_gradient_asym1_shallow_m1, times_plot_asym1_shallow_m1, '3gene_asym1_shallow_kymo_m1.png', param)\n",
    "plot_kymograph(results_asym1_shallow_m2, g_gradient_asym1_shallow_m2, times_plot_asym1_shallow_m2, '3gene_asym1_shallow_kymo_m2.png', param)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####      \n",
    "\n",
    "**1.5.2 Strong asymmetry**\n",
    "\n",
    "We now introduce a strong asymmetry by further decreasing the thresholds of repression of protein $A$ by proteins $B$ and $C$. We also decrease the thresholds of repression of protein $B$ by proteins $A$ and $C$. Again, we study the robustness of the patterns formed by Models 1 and 2 to decreases in the temporal decay of parameter $g$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "param[\"stat_K_AB\"] = 0.36\n",
    "param[\"stat_K_AC\"] = 0.4\n",
    "param[\"stat_K_BA\"] = 0.32\n",
    "param[\"stat_K_BC\"] = 0.4\n",
    "param[\"stat_K_CA\"] = 0.32\n",
    "param[\"stat_K_CB\"] = 0.36\n",
    "\n",
    "\n",
    "# Steep gradient for g\n",
    "param[\"s\"] = 2.5\n",
    "\n",
    "param[\"ng\"] = 2\n",
    "results_asym2_steep_m1, g_gradient_asym2_steep_m1, times_plot_asym2_steep_m1 = integrate(init_conc, param)\n",
    "\n",
    "param[\"ng\"] = 1\n",
    "results_asym2_steep_m2, g_gradient_asym2_steep_m2, times_plot_asym2_steep_m2 = integrate(init_conc, param)\n",
    "\n",
    "\n",
    "\n",
    "# Shallow gradient for g\n",
    "param[\"s\"] = 1.\n",
    "\n",
    "param[\"ng\"] = 2\n",
    "results_asym2_shallow_m1, g_gradient_asym2_shallow_m1, times_plot_asym2_shallow_m1 = integrate(init_conc, param)\n",
    "\n",
    "param[\"ng\"] = 1\n",
    "results_asym2_shallow_m2, g_gradient_asym2_shallow_m2, times_plot_asym2_shallow_m2 = integrate(init_conc, param)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "param[\"s\"] = 2.5\n",
    "\n",
    "plot_g_gradient(g_gradient_asym2_steep_m1, times_plot_asym2_steep_m1, '3gene_asym2_steep_kymo_g.png', param)\n",
    "plot_kymograph(results_asym2_steep_m1, g_gradient_asym2_steep_m1, times_plot_asym2_steep_m1, '3gene_asym2_steep_kymo_m1.png', param)\n",
    "plot_kymograph(results_asym2_steep_m2, g_gradient_asym2_steep_m2, times_plot_asym2_steep_m2, '3gene_asym2_steep_kymo_m2.png', param)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "param[\"s\"] = 1.\n",
    "\n",
    "plot_g_gradient(g_gradient_asym2_shallow_m1, times_plot_asym2_shallow_m1, '3gene_asym2_shallow_kymo_g.png', param)\n",
    "plot_kymograph(results_asym2_shallow_m1, g_gradient_asym2_shallow_m1, times_plot_asym2_shallow_m1, '3gene_asym2_shallow_kymo_m1.png', param)\n",
    "plot_kymograph(results_asym2_shallow_m2, g_gradient_asym2_shallow_m2, times_plot_asym2_shallow_m2, '3gene_asym2_shallow_kymo_m2.png', param)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####      \n",
    "\n",
    "**1.5.3 Random asymmetry**\n",
    "\n",
    "We now generate a random asymmetry by choosing randomly the strengths of the repression interactions of the static term. We pick the repression strengths from a Gaussian distribution of mean $0.4$ and standard deviation $0.04$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(78665433)\n",
    "random_K_asym2 = np.random.normal(loc=0.4, scale=0.04, size=(6,))\n",
    "\n",
    "param[\"stat_K_AB\"] = random_K_asym2[0]\n",
    "param[\"stat_K_AC\"] = random_K_asym2[1]\n",
    "param[\"stat_K_BA\"] = random_K_asym2[2]\n",
    "param[\"stat_K_BC\"] = random_K_asym2[3]\n",
    "param[\"stat_K_CA\"] = random_K_asym2[4]\n",
    "param[\"stat_K_CB\"] = random_K_asym2[5]\n",
    "\n",
    "\n",
    "# Steep gradient for g\n",
    "param[\"s\"] = 2.5\n",
    "\n",
    "param[\"ng\"] = 2\n",
    "results_asym3_steep_m1, g_gradient_asym3_steep_m1, times_plot_asym3_steep_m1 = integrate(init_conc, param)\n",
    "\n",
    "param[\"ng\"] = 1\n",
    "results_asym3_steep_m2, g_gradient_asym3_steep_m2, times_plot_asym3_steep_m2 = integrate(init_conc, param)\n",
    "\n",
    "\n",
    "\n",
    "# Shallow gradient for g\n",
    "param[\"s\"] = 1.\n",
    "\n",
    "param[\"ng\"] = 2\n",
    "results_asym3_shallow_m1, g_gradient_asym3_shallow_m1, times_plot_asym3_shallow_m1 = integrate(init_conc, param)\n",
    "\n",
    "param[\"ng\"] = 1\n",
    "results_asym3_shallow_m2, g_gradient_asym3_shallow_m2, times_plot_asym3_shallow_m2 = integrate(init_conc, param)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "param[\"s\"] = 2.5\n",
    "\n",
    "plot_g_gradient(g_gradient_asym3_steep_m1, times_plot_asym3_steep_m1, '3gene_asym3_steep_kymo_g.png', param)\n",
    "plot_kymograph(results_asym3_steep_m1, g_gradient_asym3_steep_m1, times_plot_asym3_steep_m1, '3gene_asym3_steep_kymo_m1.png', param)\n",
    "plot_kymograph(results_asym3_steep_m2, g_gradient_asym3_steep_m2, times_plot_asym3_steep_m2, '3gene_asym3_steep_kymo_m2.png', param)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "param[\"s\"] = 1.\n",
    "\n",
    "plot_g_gradient(g_gradient_asym3_shallow_m1, times_plot_asym3_shallow_m1, '3gene_asym3_shallow_kymo_g.png', param)\n",
    "plot_kymograph(results_asym3_shallow_m1, g_gradient_asym3_shallow_m1, times_plot_asym3_shallow_m1, '3gene_asym3_shallow_kymo_m1.png', param)\n",
    "plot_kymograph(results_asym3_shallow_m2, g_gradient_asym3_shallow_m2, times_plot_asym3_shallow_m2, '3gene_asym3_shallow_kymo_m2.png', param)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####    \n",
    "\n",
    "**2. Bifurcation diagrams**\n",
    "\n",
    "For every asymmetric 3-gene model simulated above, we generate a bifurcation diagram that shows the position (in gene expression space) of the fixed points and limit cycles as a function of parameter $g$. We plot the concentration of gene $A$ at which stable and unstable fixed points can be found, as well as the minimum and maximum concentrations of gene $A$ that are reached by the oscillations on the limit cycle. The XPP AUTO software was used to generate the data plotted on these bifurcation diagams.\n",
    "\n",
    "\n",
    "####   \n",
    "\n",
    "**2.1 Extracting the fixed points and limit cycles**\n",
    "\n",
    "We first define functions that extract the data obtained from the XPP AUTO software. The function `extract_fixed_pts` outputs a matrix containing information about the position of the fixed points in the data file and an array of colors for the fixed points, either green for a stable fixed point or red for an unstable fixed point. Similarly, the function `extract_limit_cycles` assigns a color to each limit cycle in the data file, either grey for a stable cycle or blue for an unstable cycle."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_fixed_pts(data_file):\n",
    "    \n",
    "    fixed_pts = np.genfromtxt(data_file, delimiter=' ')\n",
    "\n",
    "    fixed_pts_color = []\n",
    "    for fps in fixed_pts[:,0]:\n",
    "    \n",
    "        if (int(fps) == 1):    fixed_pts_color.append('tab:green')\n",
    "        elif (int(fps) == 2):    fixed_pts_color.append('tab:red')\n",
    "        else:\n",
    "            print('ERROR: one of the points is neither a stable nor unstable equilibria.')\n",
    "            break\n",
    "            \n",
    "    return fixed_pts, fixed_pts_color\n",
    "\n",
    "        \n",
    "\n",
    "def extract_limit_cycles(data_file):\n",
    "    \n",
    "    limit_cycles = np.genfromtxt(data_file, delimiter=' ')\n",
    "\n",
    "    limit_cycles_color = []\n",
    "    for lc in limit_cycles[:,0]:\n",
    "    \n",
    "        if (int(lc) == 3):    limit_cycles_color.append('k')\n",
    "        elif (int(lc) == 4):    limit_cycles_color.append('tab:blue')\n",
    "        else:\n",
    "            print('ERROR: one of the points is neither stable nor unstable periodics.')\n",
    "            break\n",
    "            \n",
    "    return limit_cycles, limit_cycles_color"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####   \n",
    "\n",
    "**2.2 Generating the bifurcation diagrams**\n",
    "\n",
    "We plot the bifurcation diagrams, and indicate the bifurcation events that occur in each model.\n",
    "\n",
    "\n",
    "####      \n",
    "\n",
    "**2.2.1 Weak asymmetry**\n",
    "\n",
    "We first generate the bifurcation diagrams for the 3-gene Models 1 and 2 with the simple asymmetry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model 1\n",
    "\n",
    "# Extract the data\n",
    "fp_asym1_m1, fp_asym1_m1_color = extract_fixed_pts('XPPAUTO_DATA/3gene_asym1_m1/3gene_asym1_m1_all_fp.dat')\n",
    "lc_asym1_m1, lc_asym1_m1_color = extract_limit_cycles('XPPAUTO_DATA/3gene_asym1_m1/3gene_asym1_m1_all_lc.dat')\n",
    "g_Hopf = 0.7265\n",
    "g_SN1 = 0.382\n",
    "g_SN2 = 0.35\n",
    "\n",
    "# Plot the bifurcation diagram\n",
    "plt.rcParams[\"figure.figsize\"] = 6.4, 4.2\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Limit cycles\n",
    "ax.fill_between(lc_asym1_m1[:50,3], lc_asym1_m1[:50,9], lc_asym1_m1[:50,6], color='tab:grey', alpha=0.1)\n",
    "ax.scatter(lc_asym1_m1[:,3], lc_asym1_m1[:,6], s=15, color=lc_asym1_m1_color)\n",
    "ax.scatter(lc_asym1_m1[:,3], lc_asym1_m1[:,9], s=15, color=lc_asym1_m1_color)\n",
    "\n",
    "# Fixed points\n",
    "ax.scatter(fp_asym1_m1[:,3], fp_asym1_m1[:,6], s=15, color=fp_asym1_m1_color)\n",
    "\n",
    "# Bifurcations\n",
    "ax.plot([g_Hopf, g_Hopf], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_Hopf+0.02, 0.82, 'Hopf', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN1, g_SN1], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SN1+0.02, 0.82, 'SN', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN2, g_SN2], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SN2-0.02, 0.82, 'SN', fontsize=32, horizontalalignment='left')\n",
    "    \n",
    "# Other plot options\n",
    "ax.set_ylabel('$[A]$', fontsize=32)\n",
    "ax.set_xlabel('$g$', fontsize=32)\n",
    "ax.tick_params(labelsize=24)\n",
    "ax.set_xlim([1., 0.])\n",
    "ax.set_ylim([-0.02, 1.02])\n",
    "ax.set_xticks([1., 0.5, 0.])\n",
    "ax.set_yticks([0., 0.5, 1.])\n",
    "        \n",
    "fig.tight_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model 2\n",
    "\n",
    "# Extract the data\n",
    "fp_asym1_m2, fp_asym1_m2_color = extract_fixed_pts('XPPAUTO_DATA/3gene_asym1_m2/3gene_asym1_m2_all_fp.dat')\n",
    "lc_asym1_m2, lc_asym1_m2_color = extract_limit_cycles('XPPAUTO_DATA/3gene_asym1_m2/3gene_asym1_m2_all_lc.dat')\n",
    "g_SNIC = 0.3207\n",
    "g_SN1 = 0.298\n",
    "g_SN2 = 0.27\n",
    "\n",
    "# Plot the bifurcation diagram\n",
    "plt.rcParams[\"figure.figsize\"] = 6.4, 4.2\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Limit cycles\n",
    "ax.fill_between(lc_asym1_m2[:,3], lc_asym1_m2[:,9], lc_asym1_m2[:,6], color='tab:grey', alpha=0.1)\n",
    "ax.scatter(lc_asym1_m2[:,3], lc_asym1_m2[:,6], s=15, color=lc_asym1_m2_color)\n",
    "ax.scatter(lc_asym1_m2[:,3], lc_asym1_m2[:,9], s=15, color=lc_asym1_m2_color)\n",
    "\n",
    "# Fixed points\n",
    "ax.scatter(fp_asym1_m2[:,3], fp_asym1_m2[:,6], s=15, color=fp_asym1_m2_color)\n",
    "\n",
    "# Bifurcations\n",
    "ax.plot([g_SNIC, g_SNIC], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SNIC+0.02, 0.82, 'SNIC', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN1, g_SN1], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "# ax.text(g_SN1+0.02, 0.82, 'SN', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN2, g_SN2], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SN2-0.02, 0.65, 'SN', fontsize=32, horizontalalignment='left')\n",
    "    \n",
    "# Other plot options\n",
    "ax.set_ylabel('$[A]$', fontsize=32)\n",
    "ax.set_xlabel('$g$', fontsize=32)\n",
    "ax.tick_params(labelsize=24)\n",
    "ax.set_xlim([1., 0.])\n",
    "ax.set_ylim([-0.02, 1.02])\n",
    "ax.set_xticks([1., 0.5, 0.])\n",
    "ax.set_yticks([0., 0.5, 1.])\n",
    "        \n",
    "fig.tight_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####      \n",
    "\n",
    "**2.2.2 Strong asymmetry**\n",
    "\n",
    "Bifurcation diagrams for the 3-gene Models 1 and 2 with the strong asymmetry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model 1\n",
    "\n",
    "# Extract the data\n",
    "fp_asym2_m1, fp_asym2_m1_color = extract_fixed_pts('XPPAUTO_DATA/3gene_asym2_m1/3gene_asym2_m1_all_fp.dat')\n",
    "lc_asym2_m1, lc_asym2_m1_color = extract_limit_cycles('XPPAUTO_DATA/3gene_asym2_m1/3gene_asym2_m1_all_lc.dat')\n",
    "g_Hopf = 0.742\n",
    "g_SN1 = 0.367\n",
    "g_SN2 = 0.35\n",
    "\n",
    "# Plot the bifurcation diagram\n",
    "plt.rcParams[\"figure.figsize\"] = 6.4, 4.2\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Limit cycles\n",
    "ax.fill_between(lc_asym2_m1[:50,3], lc_asym2_m1[:50,9], lc_asym2_m1[:50,6], color='tab:grey', alpha=0.1)\n",
    "ax.scatter(lc_asym2_m1[:,3], lc_asym2_m1[:,6], s=15, color=lc_asym2_m1_color)\n",
    "ax.scatter(lc_asym2_m1[:,3], lc_asym2_m1[:,9], s=15, color=lc_asym2_m1_color)\n",
    "\n",
    "# Fixed points\n",
    "ax.scatter(fp_asym2_m1[:,3], fp_asym2_m1[:,6], s=15, color=fp_asym2_m1_color)\n",
    "\n",
    "# Bifurcations\n",
    "ax.plot([g_Hopf, g_Hopf], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_Hopf+0.02, 0.82, 'Hopf', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN1, g_SN1], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SN1+0.02, 0.82, 'SN', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN2, g_SN2], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SN2-0.02, 0.82, 'SN', fontsize=32, horizontalalignment='left')\n",
    "    \n",
    "# Other plot options\n",
    "ax.set_ylabel('$[A]$', fontsize=32)\n",
    "ax.set_xlabel('$g$', fontsize=32)\n",
    "ax.tick_params(labelsize=24)\n",
    "ax.set_xlim([1., 0.])\n",
    "ax.set_ylim([-0.02, 1.02])\n",
    "ax.set_xticks([1., 0.5, 0.])\n",
    "ax.set_yticks([0., 0.5, 1.])\n",
    "        \n",
    "fig.tight_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model 2\n",
    "\n",
    "# Extract the data\n",
    "fp_asym2_m2, fp_asym2_m2_color = extract_fixed_pts('XPPAUTO_DATA/3gene_asym2_m2/3gene_asym2_m2_all_fp.dat')\n",
    "lc_asym2_m2, lc_asym2_m2_color = extract_limit_cycles('XPPAUTO_DATA/3gene_asym2_m2/3gene_asym2_m2_all_lc.dat')\n",
    "g_SNIC = 0.344\n",
    "g_SN1 = 0.26\n",
    "g_SN2 = 0.27\n",
    "\n",
    "# Plot the bifurcation diagram\n",
    "plt.rcParams[\"figure.figsize\"] = 6.4, 4.2\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Limit cycles\n",
    "ax.fill_between(lc_asym2_m2[:,3], lc_asym2_m2[:,9], lc_asym2_m2[:,6], color='tab:grey', alpha=0.1)\n",
    "ax.scatter(lc_asym2_m2[:,3], lc_asym2_m2[:,6], s=15, color=lc_asym2_m2_color)\n",
    "ax.scatter(lc_asym2_m2[:,3], lc_asym2_m2[:,9], s=15, color=lc_asym2_m2_color)\n",
    "\n",
    "# Fixed points\n",
    "ax.scatter(fp_asym2_m2[:,3], fp_asym2_m2[:,6], s=15, color=fp_asym2_m2_color)\n",
    "\n",
    "# Bifurcations\n",
    "ax.plot([g_SNIC, g_SNIC], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SNIC+0.02, 0.82, 'SNIC', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN1, g_SN1], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "# ax.text(g_SN1+0.02, 0.82, 'SN', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN2, g_SN2], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SN2-0.02, 0.65, 'SN', fontsize=32, horizontalalignment='left')\n",
    "    \n",
    "# Other plot options\n",
    "ax.set_ylabel('$[A]$', fontsize=32)\n",
    "ax.set_xlabel('$g$', fontsize=32)\n",
    "ax.tick_params(labelsize=24)\n",
    "ax.set_xlim([1., 0.])\n",
    "ax.set_ylim([-0.02, 1.02])\n",
    "ax.set_xticks([1., 0.5, 0.])\n",
    "ax.set_yticks([0., 0.5, 1.])\n",
    "        \n",
    "fig.tight_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####      \n",
    "\n",
    "**2.2.3 Random asymmetry**\n",
    "\n",
    "Bifurcation diagrams for the 3-gene Models 1 and 2 with the random asymmetry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model 1\n",
    "\n",
    "# Extract the data\n",
    "fp_asym3_m1, fp_asym3_m1_color = extract_fixed_pts('XPPAUTO_DATA/3gene_asym3_m1/3gene_asym3_m1_all_fp.dat')\n",
    "lc_asym3_m1, lc_asym3_m1_color = extract_limit_cycles('XPPAUTO_DATA/3gene_asym3_m1/3gene_asym3_m1_all_lc.dat')\n",
    "g_Hopf = 0.7265\n",
    "g_SN1 = 0.398\n",
    "g_SN2 = 0.353\n",
    "\n",
    "# Plot the bifurcation diagram\n",
    "plt.rcParams[\"figure.figsize\"] = 6.4, 4.2\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Limit cycles\n",
    "ax.fill_between(lc_asym3_m1[:50,3], lc_asym3_m1[:50,9], lc_asym3_m1[:50,6], color='tab:grey', alpha=0.1)\n",
    "ax.scatter(lc_asym3_m1[:,3], lc_asym3_m1[:,6], s=15, color=lc_asym3_m1_color)\n",
    "ax.scatter(lc_asym3_m1[:,3], lc_asym3_m1[:,9], s=15, color=lc_asym3_m1_color)\n",
    "\n",
    "# Fixed points\n",
    "ax.scatter(fp_asym3_m1[:,3], fp_asym3_m1[:,6], s=15, color=fp_asym3_m1_color)\n",
    "\n",
    "# Bifurcations\n",
    "ax.plot([g_Hopf, g_Hopf], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_Hopf+0.02, 0.82, 'Hopf', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN1, g_SN1], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SN1+0.02, 0.82, 'SN', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN2, g_SN2], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SN2-0.02, 0.82, 'SN', fontsize=32, horizontalalignment='left')\n",
    "   \n",
    "# Other plot options\n",
    "ax.set_ylabel('$[A]$', fontsize=32)\n",
    "ax.set_xlabel('$g$', fontsize=32)\n",
    "ax.tick_params(labelsize=24)\n",
    "ax.set_xlim([1., 0.])\n",
    "ax.set_ylim([-0.02, 1.02])\n",
    "ax.set_xticks([1., 0.5, 0.])\n",
    "ax.set_yticks([0., 0.5, 1.])\n",
    "        \n",
    "fig.tight_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model 2\n",
    "\n",
    "fp_asym3_m2, fp_asym3_m2_color = extract_fixed_pts('XPPAUTO_DATA/3gene_asym3_m2/3gene_asym3_m2_all_fp.dat')\n",
    "lc_asym3_m2, lc_asym3_m2_color = extract_limit_cycles('XPPAUTO_DATA/3gene_asym3_m2/3gene_asym3_m2_all_lc.dat')\n",
    "g_SNIC = 0.316\n",
    "g_SN1 = 0.31\n",
    "g_SN2 = 0.28\n",
    "\n",
    "# Plot the bifurcation diagram\n",
    "plt.rcParams[\"figure.figsize\"] = 6.4, 4.2\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Limit cycles\n",
    "ax.fill_between(lc_asym3_m2[:,3], lc_asym3_m2[:,9], lc_asym3_m2[:,6], color='tab:grey', alpha=0.1)\n",
    "ax.scatter(lc_asym3_m2[:,3], lc_asym3_m2[:,6], s=15, color=lc_asym3_m2_color)\n",
    "ax.scatter(lc_asym3_m2[:,3], lc_asym3_m2[:,9], s=15, color=lc_asym3_m2_color)\n",
    "\n",
    "# Fixed points\n",
    "ax.scatter(fp_asym3_m2[:,3], fp_asym3_m2[:,6], s=15, color=fp_asym3_m2_color)\n",
    "\n",
    "# Bifurcations\n",
    "ax.plot([g_SNIC, g_SNIC], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SNIC+0.02, 0.82, 'SNIC', fontsize=32, horizontalalignment='right')\n",
    "# ax.plot([g_SN1, g_SN1], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "# ax.text(g_SN1+0.02, 0.82, 'SN', fontsize=32, horizontalalignment='right')\n",
    "ax.plot([g_SN2, g_SN2], [-0.02, 1.02], 'k', lw=5, zorder=0)\n",
    "ax.text(g_SN2-0.02, 0.65, 'SN', fontsize=32, horizontalalignment='left')\n",
    "    \n",
    "# Other plot options\n",
    "ax.set_ylabel('$[A]$', fontsize=32)\n",
    "ax.set_xlabel('$g$', fontsize=32)\n",
    "ax.tick_params(labelsize=24)\n",
    "ax.set_xlim([1., 0.])\n",
    "ax.set_ylim([-0.02, 1.02])\n",
    "ax.set_xticks([1., 0.5, 0.])\n",
    "ax.set_yticks([0., 0.5, 1.])\n",
    "        \n",
    "fig.tight_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####    \n",
    "\n",
    "**3. Mutual information**\n",
    "\n",
    "We now plot the mutual information between the phase along the cycle of genetic oscillations at the beginning of the simulation and the region of the pattern at the end of the simulation. See the Methods and the Supplement for more details.\n",
    "\n",
    "We first define a function that extracts the data before making the graph. The data is obtained with other python codes that are included in the `MI_DATA` folder. We do not include the codes directly in this notebook, since running them can take several hours. In short, we set all cells in the same initial state at the begining of the simulation, i.e. at the same phase of the genetic oscillations. The probability of being in a given region of the pattern is calculated by computing the relative concentration of all genes at the end of the simulation. We repeat this procedure for different initial phases. Since the elongation speed is constant, we sample the initial phases uniformly on the cycle. Because the simulations are stochastic, we repeat the whole procedure several times to get reliable results. We compute and plot the mutual information for multiple values of the typical concentration $\\Omega$ and the diffusion coefficient $D$.\n",
    "\n",
    "We compute and plot the mutual information for multiple values of the steepness of the morphogen gradient. Note that the ideal mutual information is obtained by using a step function for $g$, jumping from 1 to 0. Indeed, using such a function for $g$ enforces the system to switch directly from the dynamic module to the static module. This allows us to measure the information content encoded in the static module. Note that the ideal mutual information is not $\\log(3)$ bits since the regions of the pattern are asymmetric.\n",
    "\n",
    "\n",
    "####   \n",
    "\n",
    "**3.1 Extracting the mutual information data**\n",
    "\n",
    "In the next code cell, we define a function that extracts the data of the mutual information from the `.txt` files created by the python codes. Each file has the mutual information for a given model, for one specific asymmetry and for one specific value of the morphogen gradient steepness. The function below reads all files obtained from the simulation of one model for a given asymmetry and puts it an array. Therefore, we need to specify the number of values of the steepness for which we computed the mutual information. We call this number `n_steepness`. We also need to specify the asymmetry and the model used to generate the data, via `asym_index` and `m_index`, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_mutual_info(n_steepness, asym_index, m_index):\n",
    "    \n",
    "    steepness = np.zeros(n_steepness)\n",
    "    mutual_info = np.zeros(n_steepness)\n",
    "    \n",
    "    \n",
    "    for k in range(n_steepness):\n",
    "        \n",
    "        temp_line_array = np.genfromtxt(\"MI_DATA/3gene_asym\"+str(asym_index)+\"_m\"+str(m_index)+\"/MI_data_model\"+str(m_index)+\"_s\"+str(k+1)+\".txt\", delimiter=\",\")\n",
    "        \n",
    "        steepness[k] = temp_line_array[0]\n",
    "        mutual_info[k] = temp_line_array[1]\n",
    "        \n",
    "    mutual_info_stepfct = np.genfromtxt(\"MI_DATA/3gene_asym\"+str(asym_index)+\"_m\"+str(m_index)+\"/MI_data_model\"+str(m_index)+\"_stepfct.txt\", delimiter=\",\")[1]\n",
    "    \n",
    "    \n",
    "    sort = np.argsort(steepness)\n",
    "    steepness = steepness[sort]\n",
    "    mutual_info = mutual_info[sort]\n",
    "    \n",
    "    \n",
    "    return steepness, mutual_info , mutual_info_stepfct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####   \n",
    "\n",
    "**3.2 Generating the bifurcation diagrams**\n",
    "\n",
    "We plot the mutual information as a function of the morphogen gradient steepness for both models. We generate one plot per asymmetry type.\n",
    "\n",
    "\n",
    "####      \n",
    "\n",
    "**3.2.1 Weak asymmetry**\n",
    "\n",
    "Mutual information for the 3-gene Models 1 and 2 with the weak asymmetry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Retrieve M.I. data\n",
    "s_asym1_m1, MI_asym1_m1, MI_stepfct_asym1_m1 = extract_mutual_info(36, 1, 1)\n",
    "s_asym1_m2, MI_asym1_m2, MI_stepfct_asym1_m2 = extract_mutual_info(36, 1, 2)\n",
    "\n",
    "# Generate the plot\n",
    "plt.rcParams[\"figure.figsize\"] = 7.5, 5.2\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Theoretical upper bound\n",
    "ax.plot([min(s_asym1_m1), max(s_asym1_m1)], [MI_stepfct_asym1_m1, MI_stepfct_asym1_m1], c='k', ls='-', lw=4, label='Bound')\n",
    "\n",
    "# Model 1\n",
    "ax.plot(s_asym1_m1, MI_asym1_m1, color='r', marker='o', markersize=5, lw=2, label='Model 1')\n",
    "\n",
    "# Model 2\n",
    "ax.plot(s_asym1_m2, MI_asym1_m2, color='k', marker='o', markersize=5, lw=2, label='Model 2')\n",
    "\n",
    "# Other plot options\n",
    "ax.set_xlabel('$g$ gradient steepness', fontsize=32)\n",
    "ax.set_ylabel('Mutual info. (bits)', fontsize=32)\n",
    "ax.set_xlim([0., .25])\n",
    "ax.tick_params(labelsize=24)\n",
    "ax.legend(loc=4, fontsize=24)\n",
    "\n",
    "fig.tight_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####      \n",
    "\n",
    "**3.2.2 Strong asymmetry**\n",
    "\n",
    "Mutual information for the 3-gene Models 1 and 2 with the strong asymmetry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve M.I. data\n",
    "s_asym2_m1, MI_asym2_m1, MI_stepfct_asym2_m1 = extract_mutual_info(23, 2, 1)\n",
    "s_asym2_m2, MI_asym2_m2, MI_stepfct_asym2_m2 = extract_mutual_info(33, 2, 2)\n",
    "\n",
    "\n",
    "# Generate the plot\n",
    "plt.rcParams[\"figure.figsize\"] = 7.5, 5.2\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Theoretical upper bound\n",
    "ax.plot([min(s_asym2_m1), max(s_asym2_m1)], [MI_stepfct_asym2_m1, MI_stepfct_asym2_m1], c='k', ls='-', lw=4, label='Bound')\n",
    "\n",
    "# Model 1\n",
    "ax.plot(s_asym2_m1, MI_asym2_m1, color='r', marker='o', markersize=5, lw=2, label='Model 1')\n",
    "\n",
    "# Model 2\n",
    "ax.plot(s_asym2_m2, MI_asym2_m2, color='k', marker='o', markersize=5, lw=2, label='Model 2')\n",
    "\n",
    "# Other plot options\n",
    "ax.set_xlabel('$g$ gradient steepness', fontsize=32)\n",
    "ax.set_ylabel('Mutual info. (bits)', fontsize=32)\n",
    "ax.set_xlim([0., .25])\n",
    "ax.tick_params(labelsize=24)\n",
    "ax.legend(loc=5, fontsize=24)\n",
    "\n",
    "fig.tight_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####      \n",
    "\n",
    "**3.2.3 Random asymmetry**\n",
    "\n",
    "Mutual information for the 3-gene Models 1 and 2 with the random asymmetry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve M.I. data\n",
    "s_asym3_m1, MI_asym3_m1, MI_stepfct_asym3_m1 = extract_mutual_info(36, 3, 1)\n",
    "s_asym3_m2, MI_asym3_m2, MI_stepfct_asym3_m2 = extract_mutual_info(42, 3, 2)\n",
    "\n",
    "\n",
    "# Generate the plot\n",
    "plt.rcParams[\"figure.figsize\"] = 7.5, 5.2\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Theoretical upper bound\n",
    "ax.plot([min(s_asym3_m1), max(s_asym3_m1)], [MI_stepfct_asym3_m1, MI_stepfct_asym3_m1], c='k', ls='-', lw=4, label='Bound')\n",
    "\n",
    "# Model 1\n",
    "ax.plot(s_asym3_m1, MI_asym3_m1, color='r', marker='o', markersize=5, lw=2, label='Model 1')\n",
    "\n",
    "# Model 2\n",
    "ax.plot(s_asym3_m2, MI_asym3_m2, color='k', marker='o', markersize=5, lw=2, label='Model 2')\n",
    "\n",
    "# Other plot options\n",
    "ax.set_xlabel('$g$ gradient steepness', fontsize=32)\n",
    "ax.set_ylabel('Mutual info. (bits)', fontsize=32)\n",
    "ax.set_xlim([0., .25])\n",
    "ax.tick_params(labelsize=24)\n",
    "ax.legend(loc=4, fontsize=24)\n",
    "\n",
    "fig.tight_layout()\n",
    "fig.show()"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
